### ISSUE TITLE
Re-create DD model

### ISSUE THREAD

**Issue by vhaasteren** [nanograv/PINT#39]
Currently, the DD model is a copy of the Tempo2 module DDmodel.C

_All_ these calculations need to be re-done, and re-implemented. Preferably in a way that allows easy incorporations of many of the other tempo2 binary models.

**Comment by paulray** [comment]
There are a couple of features that would be nice to have in a new implementation of binary models.

The first is to allow parameterizing of eccentric orbits as either ECC,OMEGA or EPS1,EPS2. It would be great if this choice could be made regardless of what the underlying delay model is (e.g. BT or DD), but I don't actually know how difficult that is.

The second is that it would be great if all of the binary models supported either PB,PBDOT or FB0, FB1, ... again independent of the binary delay calculation.

It would be really awesome if we could design the binary models classes to allow for this with a minimal amount of code duplication.

**Comment by vhaasteren** [comment]
We would have to define a 'mapping' or 'transformation' interface, that maps one parameterization into another. It would just be a transformation of coordinates, where the derivatives would also require the jacobian transformation.

It would be best if we can define these transformation classes separately. They typically only transform a subset of parameters, and we could potentially have several of those in a single timing model. These transformation objects could then be attributes of the timing model class.

Only problem I see with that is that the TimingModel class initializes the .params list at the time of __init__. So all the binary models have defined their parameters at the class level. Ideally, you would use the same binary model class, and apply a transformation to it....

**Comment by luojing1211** [comment]
As I said on the telecom, I will focus the DD model next. Let me understand
this. Are we going to build a binary(TimingModel) class to kind have a
general control of different binary model?

-L

**Comment by vhaasteren** [comment]
Best way to start would probably be to do 'naked' tests of your binary model, and compare to libstempo values for the binary delay. I'll show you how I have been doing that for the BT model in an IPython notebook (will send you right away).

Only once you have that working well enough would I recommend you implementing it in the PINT timing model class. Added extra advantage is that, if implemented separately as a PINT package, people can import the PINT package of binary models, without having to run on datasets.

**Comment by luojing1211** [comment]
We have dd model now. This should be closed. If need more optimise on dd model, let us open a new one.
